validator/sawtooth_validator/gossip/gossip_handlers.py

```diff
@@ -233,32 +232,40 @@ def __init__(self, gossip, completer):
         self._gossip = gossip
         self._completer = completer
 
-    def handle(self, connection_id, message_content):
+    # pylint:disable=arguments-differ
+    def handle(self, connection_id, obj_tag_ttl):
+        obj, tag, ttl = obj_tag_ttl
+
         exclude = [connection_id]
-        gossip_message = GossipMessage()
-        gossip_message.ParseFromString(message_content)
-        if gossip_message.time_to_live == 0:
+
+        if ttl <= 0:
             # Do not forward message if it has reached its time to live limit
             return HandlerResult(status=HandlerStatus.PASS)
 
-        else:
-            # decrement time_to_live
-            time_to_live = gossip_message.time_to_live
-            gossip_message.time_to_live = time_to_live - 1
-
-        if gossip_message.content_type == GossipMessage.BATCH:
-            batch = Batch()
-            batch.ParseFromString(gossip_message.content)
+        if tag == GossipMessage.BATCH:
             # If we already have this batch, don't forward it
-            if not self._completer.get_batch(batch.header_signature):
-                self._gossip.broadcast_batch(batch, exclude)
-        elif gossip_message.content_type == GossipMessage.BLOCK:
-            block = Block()
-            block.ParseFromString(gossip_message.content)
+            if not self._completer.get_batch(obj.header_signature):
+                self._gossip.broadcast_batch(obj, exclude)
+        elif tag == GossipMessage.BLOCK:
             # If we already have this block, don't forward it
-            if not self._completer.get_block(block.header_signature):
-                self._gossip.broadcast_block(block, exclude)
+            if not self._completer.get_block(obj.header_signature):
+                self._gossip.broadcast_block(obj, exclude)
         else:
-            LOGGER.info("received %s, not BATCH or BLOCK",
-                        gossip_message.content_type)
+            LOGGER.info("received %s, not BATCH or BLOCK", tag)
         return HandlerResult(status=HandlerStatus.PASS)
+
+
+def gossip_message_preprocessor(message_content_bytes):
+    gossip_message = GossipMessage()
+    gossip_message.ParseFromString(message_content_bytes)
+
+    tag = gossip_message.content_type
+
+    if tag == GossipMessage.BLOCK:
+        obj = Block()
+        obj.ParseFromString(gossip_message.content)
+    elif tag == GossipMessage.BATCH:
+        obj = Batch()
+        obj.ParseFromString(gossip_message.content)
+
+    return obj, tag, gossip_message.time_to_live - 1
```

Comment:
Time to live should not be reduced by the preprocessor - these changes should be made by the handlers, not during what is effectively a parsing phase.